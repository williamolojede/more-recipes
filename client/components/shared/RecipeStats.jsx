/* eslint react/require-default-props: 0 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';

class RecipeStats extends Component {
  constructor() {
    super();
    this.state = {
      activeButton: ''
    };
  }

  onButtonClick = (type, direction) => {
    if (type === 'favorite') {
      this.setState({
        activeButton: 'favoriteButton'
      });
      this.props.favorite();
    }

    if (type === 'vote') {
      this.setState({
        activeButton: `${direction}VoteButton`
      });
      this.props.vote(direction);
    }
  }
  render() {
    const {
      favoriteCount,
      upVoteCount,
      downVoteCount,
      viewCount
    } = this.props.stats;
    return (
      <ul className="recipe-stats">
        <li className="recipe-stats__favorite">
          <button
            onClick={() => this.props.favorite && this.onButtonClick('favorite')}
            className={this.state.activeButton === 'favoriteButton' && 'active'}
          >
            <i className="mdi mdi-heart" />
            <span>{favoriteCount}</span>
          </button>
        </li>
        <li className="recipe-stats__vote-up">
          <button
            onClick={() => this.props.vote && this.onButtonClick('vote', 'up')}
            className={this.state.activeButton === 'upVoteButton' && 'active'}
          >
            <i className="mdi mdi-thumb-up" />
            <span>{upVoteCount}</span>
          </button>
        </li>
        <li className="recipe-stats__vote-down">
          <button
            onClick={() => this.props.vote && this.onButtonClick('vote', 'down')}
            className={this.state.activeButton === 'downVoteButton' && 'active'}
          >
            <i className="mdi mdi-thumb-down" />
            <span>{downVoteCount}</span>
          </button>
        </li>
        <li className="recipe-stats__views">
          <i className="mdi mdi-eye" />
          <span>{viewCount}</span>
        </li>
      </ul>
    );
  }
}

RecipeStats.propTypes = {
  stats: PropTypes.shape({
    downVoteCount: PropTypes.number.isRequired,
    favoriteCount: PropTypes.number.isRequired,
    upVoteCount: PropTypes.number.isRequired,
    viewCount: PropTypes.number.isRequired
  }).isRequired,

  vote: PropTypes.func,
  favorite: PropTypes.func
};


export default RecipeStats;
