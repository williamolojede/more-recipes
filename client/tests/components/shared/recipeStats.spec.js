import React from 'react';
import renderer from 'react-test-renderer';
import RecipeStats from '../../../components/shared/RecipeStats';

function setup() {
  const props = {
    stats: {
      downVoteCount: 0,
      favoriteCount: 0,
      upVoteCount: 0,
      viewCount: 0
    },
    favorite: jest.fn(),
    vote: jest.fn()
  };

  const component = renderer.create(<RecipeStats {...props} />);
  const wrapper = shallow(<RecipeStats {...props} />);
  return {
    component,
    wrapper
  };
}

describe('Recipe Stats', () => {
  it('should render all expected elements', () => {
    const tree = setup().component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should favorite as expected', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
    const button = wrapper.find('.recipe-stats__favorite button');
    button.simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
  it('should up vote as expected', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
    const button = wrapper.find('.recipe-stats__vote-up button');
    button.simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
  it('should downvote as expected', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
    const button = wrapper.find('.recipe-stats__vote-down button');
    button.simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
});
