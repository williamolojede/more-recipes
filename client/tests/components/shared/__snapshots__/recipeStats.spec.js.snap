// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeStats should downvote as expected 1`] = `
<ul
  className="recipe-stats"
>
  <li
    className="recipe-stats__favorite"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-heart"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-up"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-up"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-down"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-down"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__views"
  >
    <i
      className="mdi mdi-eye"
    />
    <span>
      0
    </span>
  </li>
</ul>
`;

exports[`RecipeStats should downvote as expected 2`] = `
<ul
  className="recipe-stats"
>
  <li
    className="recipe-stats__favorite"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-heart"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-up"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-up"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-down"
  >
    <button
      className="active"
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-down"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__views"
  >
    <i
      className="mdi mdi-eye"
    />
    <span>
      0
    </span>
  </li>
</ul>
`;

exports[`RecipeStats should favorite as expected 1`] = `
<ul
  className="recipe-stats"
>
  <li
    className="recipe-stats__favorite"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-heart"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-up"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-up"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-down"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-down"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__views"
  >
    <i
      className="mdi mdi-eye"
    />
    <span>
      0
    </span>
  </li>
</ul>
`;

exports[`RecipeStats should favorite as expected 2`] = `
<ul
  className="recipe-stats"
>
  <li
    className="recipe-stats__favorite"
  >
    <button
      className="active"
      onClick={[Function]}
    >
      <i
        className="mdi mdi-heart"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-up"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-up"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-down"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-down"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__views"
  >
    <i
      className="mdi mdi-eye"
    />
    <span>
      0
    </span>
  </li>
</ul>
`;

exports[`RecipeStats should render all expected elements 1`] = `
<ul
  className="recipe-stats"
>
  <li
    className="recipe-stats__favorite"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-heart"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-up"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-up"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-down"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-down"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__views"
  >
    <i
      className="mdi mdi-eye"
    />
    <span>
      0
    </span>
  </li>
</ul>
`;

exports[`RecipeStats should up vote as expected 1`] = `
<ul
  className="recipe-stats"
>
  <li
    className="recipe-stats__favorite"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-heart"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-up"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-up"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-down"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-down"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__views"
  >
    <i
      className="mdi mdi-eye"
    />
    <span>
      0
    </span>
  </li>
</ul>
`;

exports[`RecipeStats should up vote as expected 2`] = `
<ul
  className="recipe-stats"
>
  <li
    className="recipe-stats__favorite"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-heart"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-up"
  >
    <button
      className="active"
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-up"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__vote-down"
  >
    <button
      className={false}
      onClick={[Function]}
    >
      <i
        className="mdi mdi-thumb-down"
      />
      <span>
        0
      </span>
    </button>
  </li>
  <li
    className="recipe-stats__views"
  >
    <i
      className="mdi mdi-eye"
    />
    <span>
      0
    </span>
  </li>
</ul>
`;
